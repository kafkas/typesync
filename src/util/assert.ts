class AssertionError extends Error {
  public constructor(message: string) {
    super(message);
  }
}

export function assert(condition: unknown, message?: string): asserts condition {
  if (!condition) {
    throw new AssertionError(message ?? 'Assertion failed');
  }
}

export function assertNever(val: never): never {
  throw new AssertionError(`Invalid value: ${JSON.stringify(val)}`);
}

export function assertNeverNoThrow(_val: never): void {}

export function assertEmpty<T extends true | false>(_t: T) {}
