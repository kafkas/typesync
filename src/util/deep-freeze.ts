export function deepFreeze<T>(input: T) {
  if (Array.isArray(input)) {
    for (const elem of input) {
      const unknownElem = elem as unknown;
      deepFreeze(unknownElem);
    }
  } else if (typeof input === 'object' && input !== null) {
    const propNames = Object.getOwnPropertyNames(input);

    for (const name of propNames) {
      if (name in input) {
        const value = (input as { [K: string]: unknown })[name];
        deepFreeze(value);
      }
    }
  }

  return Object.freeze(input);
}
