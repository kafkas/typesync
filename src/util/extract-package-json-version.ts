import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { z } from 'zod';

import { getDirName } from './fs.js';

export function extractPackageJsonVersion() {
  const packageJsonSchema = z.object({
    version: z.string(),
  });
  const dirName = getDirName(import.meta.url);
  const pathToPackageJson = resolve(dirName, '../../package.json');
  const packageJsonRaw = JSON.parse(readFileSync(pathToPackageJson).toString());
  const { version } = packageJsonSchema.parse(packageJsonRaw);
  return version;
}
