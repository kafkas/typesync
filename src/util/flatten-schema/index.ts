export { flattenSchema } from './flatten-schema';
