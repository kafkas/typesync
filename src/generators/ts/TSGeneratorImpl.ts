import { StringBuilder } from '@proficient/ds';

import { createGenerationOutput } from '../../implementations/GenerationOutputImpl';
import type { Generator, TSGeneratorConfig } from '../../interfaces';
import { schema } from '../../schema';
import { assertNever } from '../../util/assert';
import { space } from '../../util/space';

class TSGeneratorImpl implements Generator {
  private get firestore() {
    switch (this.config.platform) {
      case 'ts:firebase-admin:11':
        return 'firestore';
      default:
        assertNever(this.config.platform);
    }
  }

  public constructor(private readonly config: TSGeneratorConfig) {}

  public async generate(s: schema.Schema) {
    const { models } = s;

    const builder = new StringBuilder();

    const tsFirestoreImport = this.getImportFirestoreStatement();

    builder.append(`${tsFirestoreImport}\n\n`);

    const { aliasModels, documentModels } = this.divideModelsByType(models);

    aliasModels.forEach(model => {
      const tsType = this.getTSTypeForAliasModel(model, 0);
      if (model.docs !== undefined) {
        const tsDoc = this.buildTSDoc(model.docs, 0);
        builder.append(`${tsDoc}\n`);
      }
      builder.append(`export type ${model.name} = ${tsType};\n\n`);
    });

    documentModels.forEach((model, modelIndex) => {
      // A Firestore document can be considered a 'map' type
      const tsType = this.getTSTypeForMapValueType({ type: 'map', fields: model.fields }, 0);
      if (model.docs !== undefined) {
        const tsDoc = this.buildTSDoc(model.docs, 0);
        builder.append(`${tsDoc}\n`);
      }
      builder.append(`export interface ${model.name} ${tsType}\n`);
      if (modelIndex < documentModels.length - 1) {
        builder.append('\n');
      }
    });

    return createGenerationOutput(builder.toString());
  }

  private divideModelsByType(models: schema.Model[]) {
    const aliasModels: schema.AliasModel[] = [];
    const documentModels: schema.DocumentModel[] = [];
    models.forEach(model => {
      switch (model.type) {
        case 'alias':
          aliasModels.push(model);
          break;
        case 'document':
          documentModels.push(model);
          break;
        default:
          assertNever(model);
      }
    });
    return { aliasModels, documentModels };
  }

  private getImportFirestoreStatement() {
    switch (this.config.platform) {
      case 'ts:firebase-admin:11':
        return `import { firestore } from 'firebase-admin';`;
      default:
        assertNever(this.config.platform);
    }
  }

  private getTSTypeForAliasModel(model: schema.AliasModel, depth: number) {
    return this.getTSTypeForValueType(model.value, depth);
  }

  private getTSTypeForValueType(type: schema.types.Type, depth: number) {
    if (schema.isPrimitiveValueType(type)) {
      return this.getTSTypeForPrimitiveValueType(type);
    }
    switch (type.type) {
      case 'literal':
        return this.getTSTypeForLiteralValueType(type);
      case 'enum':
        return this.getTSTypeForEnumValueType(type);
      case 'tuple':
        return this.getTSTypeForTupleValueType(type, depth);
      case 'list':
        return this.getTSTypeForListValueType(type, depth);
      case 'map':
        return this.getTSTypeForMapValueType(type, depth);
      case 'union':
        return this.getTSTypeForUnionValueType(type, depth);
      case 'alias':
        return type.name;
      default:
        assertNever(type);
    }
  }

  private getTSTypeForPrimitiveValueType(type: schema.types.Primitive) {
    switch (type.type) {
      case 'nil':
        return 'null';
      case 'string':
        return 'string';
      case 'boolean':
        return 'boolean';
      case 'int':
        return 'number';
      case 'timestamp':
        return `${this.firestore}.Timestamp`;
      default:
        assertNever(type.type);
    }
  }

  private getTSTypeForLiteralValueType(type: schema.types.Literal) {
    switch (typeof type.value) {
      case 'string':
        return `'${type.value}'`;
      case 'number':
        return `${type.value}`;
      case 'boolean':
        return `${type.value}`;
      default:
        assertNever(type.value);
    }
  }

  private getTSTypeForEnumValueType(type: schema.types.Enum) {
    const { items } = type;
    return items
      .map(({ value }) => {
        switch (typeof value) {
          case 'string':
            return `'${value}'`;
          case 'number':
            return `${value}`;
          default:
            assertNever(value);
        }
      })
      .join(' | ');
  }

  private getTSTypeForTupleValueType(type: schema.types.Tuple, depth: number): string {
    const tsTypes = type.values.map(v => this.getTSTypeForValueType(v, depth)).join(', ');
    return `[${tsTypes}]`;
  }

  private getTSTypeForListValueType(type: schema.types.List, depth: number): string {
    const tsType = this.getTSTypeForValueType(type.of, depth);
    return `${tsType}[]`;
  }

  private getTSTypeForMapValueType(type: schema.types.Map, depth: number) {
    const { fields } = type;
    const builder = new StringBuilder();

    builder.append(`{\n`);
    fields.forEach(field => {
      if (field.docs !== undefined) {
        const tsDoc = this.buildTSDoc(field.docs, depth + 1);
        builder.append(`${tsDoc}\n`);
      }

      builder.append(space(this.config.indentation * (1 + depth)));

      const tsType = this.getTSTypeForValueType(field.type, depth + 1);
      builder.append(`${field.name}${field.optional ? '?' : ''}: ${tsType};\n`);
    });

    builder.append(space(this.config.indentation * depth));
    builder.append(`}`);

    return builder.toString();
  }

  private getTSTypeForUnionValueType(type: schema.types.Union, depth: number) {
    const tsTypes: string[] = type.members.map(memberValueType => {
      return this.getTSTypeForValueType(memberValueType, depth);
    });

    return tsTypes.join(' | ');
  }

  private buildTSDoc(docs: string, depth: number) {
    const spaces = space(this.config.indentation * depth);
    return `${spaces}/**\n${spaces} * ${docs}\n${spaces} */`;
  }
}

export function createTSGenerator(config: TSGeneratorConfig): Generator {
  return new TSGeneratorImpl(config);
}
