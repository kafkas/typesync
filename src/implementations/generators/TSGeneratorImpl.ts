import { StringBuilder } from '@proficient/ds';

import type {
  Generator,
  Schema,
  SchemaDocumentModel,
  SchemaAliasModel,
  SchemaModel,
  SchemaValueType,
  SchemaMapValueType,
  SchemaEnumValueType,
  TSGeneratorConfig,
  SchemaUnionValueType,
} from '../../interfaces';
import { createGenerationOutput } from '../GenerationOutputImpl';
import { assertNever } from '../../util/assert';

export class TSGeneratorImpl implements Generator {
  private get firestore() {
    switch (this.config.platform) {
      case 'ts:firebase-admin:11':
        return 'firestore';
      default:
        assertNever(this.config.platform);
    }
  }

  public constructor(private readonly config: TSGeneratorConfig) {}

  public async generate(schema: Schema) {
    const { models } = schema;

    const builder = new StringBuilder();

    const tsFirestoreImport = this.getImportFirestoreStatement();

    builder.append(`${tsFirestoreImport}\n\n`);

    const { aliasModels, documentModels } = this.divideModelsByType(models);

    aliasModels.forEach(model => {
      const tsType = this.getTSTypeForAliasModel(model, 0);
      if (model.docs !== undefined) {
        const tsDoc = this.buildTSDoc(model.docs, 0);
        builder.append(`${tsDoc}\n`);
      }
      builder.append(`export type ${model.name} = ${tsType};\n`);
    });

    documentModels.forEach(model => {
      // A Firestore document can be considered a 'map' type
      const tsType = this.getTSTypeForSchemaMapValueType({ type: 'map', fields: model.fields }, 0);
      if (model.docs !== undefined) {
        const tsDoc = this.buildTSDoc(model.docs, 0);
        builder.append(`${tsDoc}\n`);
      }
      builder.append(`export interface ${model.name} ${tsType}\n`);
    });

    return createGenerationOutput(builder.toString());
  }

  private getImportFirestoreStatement() {
    switch (this.config.platform) {
      case 'ts:firebase-admin:11':
        return `import { firestore } from 'firebase-admin';`;
      default:
        assertNever(this.config.platform);
    }
  }

  private divideModelsByType(models: SchemaModel[]) {
    const aliasModels: SchemaAliasModel[] = [];
    const documentModels: SchemaDocumentModel[] = [];
    models.forEach(model => {
      switch (model.type) {
        case 'alias':
          aliasModels.push(model);
          break;
        case 'document':
          documentModels.push(model);
          break;
        default:
          assertNever(model);
      }
    });
    return { aliasModels, documentModels };
  }

  private getTSTypeForAliasModel(model: SchemaAliasModel, depth: number) {
    return this.getTSTypeForSchemaValueType(model.value, depth);
  }

  private buildTSDoc(docs: string, depth: number) {
    const spaceCount = this.config.indentation * depth;
    const spaces = new Array<string>(spaceCount).fill(' ').join('');
    return `${spaces}/**\n${spaces} * ${docs}\n${spaces} */`;
  }

  private getTSTypeForSchemaValueType(type: SchemaValueType, depth: number) {
    switch (type.type) {
      case 'nil':
        return 'null';
      case 'string':
        return 'string';
      case 'boolean':
        return 'boolean';
      case 'int':
        return 'number';
      case 'timestamp':
        return `${this.firestore}.Timestamp`;
      case 'alias':
        return type.name;
      case 'enum':
        return this.getTSTypeForSchemaEnumValueType(type);
      case 'map':
        return this.getTSTypeForSchemaMapValueType(type, depth);
      case 'union':
        return this.getTSTypeForSchemaUnionValueType(type, depth);
      default:
        assertNever(type);
    }
  }

  private getTSTypeForSchemaEnumValueType(type: SchemaEnumValueType) {
    const { items } = type;
    return items
      .map(({ value }) => {
        switch (typeof value) {
          case 'string':
            return `'${value}'`;
          case 'number':
            return `${value}`;
          default:
            assertNever(value);
        }
      })
      .join(' | ');
  }

  private getTSTypeForSchemaMapValueType(type: SchemaMapValueType, depth: number) {
    const { fields } = type;
    const builder = new StringBuilder();

    builder.append(`{\n`);
    fields.forEach(field => {
      if (field.docs !== undefined) {
        const tsDoc = this.buildTSDoc(field.docs, depth + 1);
        builder.append(`${tsDoc}\n`);
      }

      builder.append('  ');

      const spaceCount = this.config.indentation * depth;
      const spaces = new Array<string>(spaceCount).fill(' ').join('');
      builder.append(spaces);

      const tsType = this.getTSTypeForSchemaValueType(field.type, depth + 1);
      builder.append(`${field.name}${field.optional ? '?' : ''}: ${tsType};\n`);
    });

    const spaceCount = this.config.indentation * depth;
    const spaces = new Array<string>(spaceCount).fill(' ').join('');
    builder.append(spaces);
    builder.append(`}`);

    return builder.toString();
  }

  private getTSTypeForSchemaUnionValueType(type: SchemaUnionValueType, depth: number) {
    const tsTypes: string[] = type.members.map(memberValueType => {
      return this.getTSTypeForSchemaValueType(memberValueType, depth);
    });

    return tsTypes.join(' | ');
  }
}

export function createTSGenerator(config: TSGeneratorConfig): Generator {
  return new TSGeneratorImpl(config);
}
