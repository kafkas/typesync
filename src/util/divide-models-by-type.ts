import type { schema } from '../interfaces';
import { assertNever } from './assert';

export function divideModelsByType(models: schema.Model[]) {
  const aliasModels: schema.AliasModel[] = [];
  const documentModels: schema.DocumentModel[] = [];
  models.forEach(model => {
    switch (model.type) {
      case 'alias':
        aliasModels.push(model);
        break;
      case 'document':
        documentModels.push(model);
        break;
      default:
        assertNever(model);
    }
  });
  return { aliasModels, documentModels };
}
