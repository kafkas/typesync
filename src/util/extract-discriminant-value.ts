import type { schema as schemaNew } from '../schema-new/index.js';
import { assert } from './assert.js';

export function extractDiscriminantValueNew(
  union: schemaNew.types.DiscriminatedUnion,
  variant: schemaNew.types.Object
) {
  const discriminantField = variant.fields.find(field => field.name === union.discriminant);
  assert(
    discriminantField && discriminantField.type.type === 'string-literal',
    'Expected object to contain a discriminant field.'
  );
  return discriminantField.type.value;
}
