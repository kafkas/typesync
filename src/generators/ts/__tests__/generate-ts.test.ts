import { loadSchemaForTestDefinition } from '../../../../test/util/load-schema';
import { createTSGenerator } from '../TSGeneratorImpl';

describe('TSGeneratorImpl', () => {
  it('generates the correct output for flat schema', async () => {
    const generator = createTSGenerator({
      indentation: 4,
      platform: 'ts:firebase-admin:11',
    });
    const schema = loadSchemaForTestDefinition('flat');
    const output = await generator.generate(schema);
    expect(output.toString()).toMatchSnapshot();
  });
});
