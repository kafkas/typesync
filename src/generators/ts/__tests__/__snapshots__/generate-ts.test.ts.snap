// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TSGeneratorImpl produces the correct generation for a flat schema 1`] = `
{
  "declarations": [
    {
      "modelName": "Username",
      "modelType": {
        "type": "string",
      },
      "type": "alias",
    },
    {
      "modelName": "UserRole",
      "modelType": {
        "items": [
          {
            "label": "ADMIN",
            "value": "admin",
          },
          {
            "label": "USER",
            "value": "user",
          },
          {
            "label": "GUEST",
            "value": "guest",
          },
        ],
        "type": "enum",
      },
      "type": "alias",
    },
    {
      "modelName": "Address",
      "modelType": {
        "properties": [
          {
            "docs": undefined,
            "name": "street",
            "optional": false,
            "type": {
              "type": "string",
            },
          },
          {
            "docs": undefined,
            "name": "city",
            "optional": false,
            "type": {
              "type": "string",
            },
          },
          {
            "docs": undefined,
            "name": "zip_code",
            "optional": false,
            "type": {
              "type": "string",
            },
          },
        ],
        "type": "object",
      },
      "type": "alias",
    },
    {
      "modelName": "Cat",
      "modelType": {
        "properties": [
          {
            "docs": undefined,
            "name": "type",
            "optional": false,
            "type": {
              "type": "literal",
              "value": "cat",
            },
          },
          {
            "docs": undefined,
            "name": "name",
            "optional": false,
            "type": {
              "type": "string",
            },
          },
          {
            "docs": undefined,
            "name": "lives_left",
            "optional": false,
            "type": {
              "type": "number",
            },
          },
        ],
        "type": "object",
      },
      "type": "alias",
    },
    {
      "modelName": "Dog",
      "modelType": {
        "properties": [
          {
            "docs": undefined,
            "name": "type",
            "optional": false,
            "type": {
              "type": "literal",
              "value": "dog",
            },
          },
          {
            "docs": undefined,
            "name": "name",
            "optional": false,
            "type": {
              "type": "string",
            },
          },
          {
            "docs": undefined,
            "name": "breed",
            "optional": false,
            "type": {
              "type": "string",
            },
          },
        ],
        "type": "object",
      },
      "type": "alias",
    },
    {
      "modelName": "Pet",
      "modelType": {
        "members": [
          {
            "name": "Cat",
            "type": "alias",
          },
          {
            "name": "Dog",
            "type": "alias",
          },
        ],
        "type": "union",
      },
      "type": "alias",
    },
    {
      "modelName": "UserProfile",
      "modelType": {
        "properties": [
          {
            "docs": undefined,
            "name": "username",
            "optional": false,
            "type": {
              "name": "Username",
              "type": "alias",
            },
          },
          {
            "docs": undefined,
            "name": "age",
            "optional": false,
            "type": {
              "type": "number",
            },
          },
          {
            "docs": undefined,
            "name": "is_active",
            "optional": false,
            "type": {
              "type": "boolean",
            },
          },
          {
            "docs": undefined,
            "name": "role",
            "optional": false,
            "type": {
              "name": "UserRole",
              "type": "alias",
            },
          },
          {
            "docs": undefined,
            "name": "created_at",
            "optional": true,
            "type": {
              "type": "timestamp",
            },
          },
          {
            "docs": undefined,
            "name": "location",
            "optional": false,
            "type": {
              "type": "tuple",
              "values": [
                {
                  "type": "number",
                },
                {
                  "type": "number",
                },
              ],
            },
          },
          {
            "docs": undefined,
            "name": "address",
            "optional": false,
            "type": {
              "name": "Address",
              "type": "alias",
            },
          },
          {
            "docs": undefined,
            "name": "favorite_numbers",
            "optional": false,
            "type": {
              "of": {
                "type": "number",
              },
              "type": "list",
            },
          },
          {
            "docs": undefined,
            "name": "pets",
            "optional": false,
            "type": {
              "of": {
                "name": "Pet",
                "type": "alias",
              },
              "type": "list",
            },
          },
          {
            "docs": undefined,
            "name": "bio",
            "optional": false,
            "type": {
              "members": [
                {
                  "type": "null",
                },
                {
                  "type": "string",
                },
              ],
              "type": "union",
            },
          },
          {
            "docs": undefined,
            "name": "area_code",
            "optional": false,
            "type": {
              "type": "literal",
              "value": 34,
            },
          },
        ],
        "type": "object",
      },
      "type": "interface",
    },
  ],
  "type": "ts",
}
`;
